tree/dataframe/inc/ROOT/RDF/RMergeableValue.hxx

```diff
@@ -0,0 +1,458 @@
+/**
+ \file ROOT/RDF/RMergeableValue.hxx
+ \ingroup dataframe
+ \author Vincenzo Eduardo Padulano
+ \author Enrico Guiraud
+ \date 2020-05
+*/
+
+/*************************************************************************
+ * Copyright (C) 1995-2020, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT_RDF_RMERGEABLEVALUE
+#define ROOT_RDF_RMERGEABLEVALUE
+
+#include <functional>
+#include <memory>
+#include <stdexcept>
+
+#include "RtypesCore.h"
+#include "TList.h" // RMergeableFill::Merge
+
+namespace ROOT {
+namespace Detail {
+namespace RDF {
+
+// Fwd declarations for RMergeableValue
+template <typename T>
+class RMergeableValue;
+
+template <typename T, typename... Ts>
+std::unique_ptr<RMergeableValue<T>> MergeValues(std::unique_ptr<RMergeableValue<T>> OutputMergeable,
```

Comment:
Design comment: the fact of having `template <typename T, typename... Ts>` means we can potentially instantiate this template for multiple types, which can lead to a merge incompatibility at runtime. Is there any way we can enforce all the `T`s to be the same at compile time or we need a runtime check?

Reply:
Yes indeed a `static_assert` could be inserted in the function body.
I have though about it, but I have to say that I personally see no instance where `MergeValues` could try to merge two incompatible values without throwing erros. For instance two notable cases may be:
* Merging Histo1D with Graph: they respectively are wrapped by `RMergeableFill<TH1D>` and `RMergeableFill<TGraph>` and when the function calls `RMergeableFill<TH1D>::Merge(const RMergeableValue<TH1D> &)` the argument has `TGraph` in the template and that produces a compilation error
* Merging Min and Max: they both are wrapped by `RMergeableGenericAction<double>` in the Python side and when the function calls `RMergeableGenericAction<double>::Merge(const RMergeableValue<double> &)` the check for std::bad_cast present in the Merge method will throw a std::invalid_argument at runtime

So in the end some error should be thrown anyway if a user tries to merge incompatible types. That said, if it's desirable to have an early stopper even before trying to call `Merge` on the mergeables, I will add the `static_assert`

Reply:
static assert added in the latest commits